app/models/concerns/orchestration/external_ipam.rb

```diff
@@ -0,0 +1,145 @@
+  end
+
+  def del_add_external_ip(params)
+    new_record? ? rollback_on_add : rollback_on_update
+  end
+
+  def del_remove_external_ip(params)
+  end
+
+  def rollback_on_add
+    if ipv4_errors_not_ipv6? && !ip_is_available?(self.ip6, self.subnet6)
+      subnet.external_ipam_proxy.delete_ip_from_subnet(self.ip6, self.subnet6.network_address)
+    elsif ipv6_errors_not_ipv4? && !ip_is_available?(self.ip, self.subnet)
+      subnet.external_ipam_proxy.delete_ip_from_subnet(self.ip, self.subnet.network_address)
+    end
+  end
+
+  def rollback_on_update
+    if ipv4_errors_not_ipv6? && ip_is_available?(old.ip, old.subnet)
+      subnet.external_ipam_proxy.add_ip_to_subnet(old.ip, old.subnet.network_address)
+    elsif ipv6_errors_not_ipv4? && ip_is_available?(old.ip6, old.subnet6)
+      subnet.external_ipam_proxy.add_ip_to_subnet(old.ip6, old.subnet6.network_address)
+    end
+  end
+
+  def requires_update?
+    return false if new_record?
+    old.ip != self.ip
+  end
+
+  def requires_update6?
+    return false if new_record?
+    old.ip6 != self.ip6
+  end
+
+  def requires_delete?
+    old = Nic::Base.find(id)
+    !old.ip.nil? && !old.subnet_id.nil?
+  end
+
+  def requires_delete6?
+    old = Nic::Base.find(id)
+    !old.ip6.nil? && !old.subnet6_id.nil?
+  end
+
+  def ip_is_available?(ip, subnet)
+    !subnet.external_ipam_proxy.ip_exists(ip, subnet.network_address)
+  end
+
+  def ipv4_errors_not_ipv6?
+    !self.errors[:ip].empty? && self.errors[:ip6].empty? && self.ip6.present?
+  end
+
+  def ipv6_errors_not_ipv4?
+    !self.errors[:ip6].empty? && self.errors[:ip].empty? && self.ip.present?
+  end
+
+  def ipv4_removed?
+    old.ip.present? && ip.blank?
+  end
+
+  def ipv6_removed?
+    old.ip6.present? && ip6.blank?
+  end
+
+  def ipv4_added?
+    old.ip.blank? && ip.present?
+  end
+
+  def ipv6_added?
+    old.ip6.blank? && ip6.present?
+  end
+
+  def ipv4_changed?
+    old.ip.present? && ip.present? && old.ip != self.ip
+  end
+
+  def ipv6_changed?
+    old.ip6.present? && ip6.present? && old.ip6 != self.ip6
+  end
+
+  private
+
+  def queue_external_ipam
+    new_record? ? queue_external_ipam_create : queue_external_ipam_update
+  end
+
+  def queue_external_ipam_create
+    return unless (external_ipam?(subnet) || external_ipam?(subnet6)) && errors.empty?
+    logger.debug "Scheduling new IP reservation(s) in external IPAM for #{self}"
+    queue.create(id: generate_external_ipam_task_id("create", subnet.network_type), name: _("Creating IP in External IPAM for %s") % self, priority: 10, action: [self, :set_add_external_ip, {:ip => ip, :subnet => subnet}]) if ip.present? && subnet.present? && external_ipam?(subnet)
+    queue.create(id: generate_external_ipam_task_id("create", subnet6.network_type), name: _("Creating IP in External IPAM for %s") % self, priority: 10, action: [self, :set_add_external_ip, {:ip => ip6, :subnet => subnet6}]) if ip6.present? && subnet6.present? && external_ipam?(subnet6)
```

Comment:
Nitpick: Change IP to IPv4 or IPv6 respectively in the task name so we can tell when error appears what broke. I am not sure if task ID is shown in errors.